### PR TITLE
Cleanup options

### PR DESCRIPTION
**Pull request by gmcgibbon**
Add ability to pass options to cleanup/clear methods.

Also protect methods that should be protected in base storage subclasses.

### CODE REVIEW

**Comment by gmcgibbon** [comment]
> Nit: I would consider breaking this into two PR's (one for supporting cleanup options, and another for protecting protected methods), but it's totally not a big deal.

I'll be merging this as two commits without squashing so the history is essentially the same as 2 PRs.

> Also, just out of curiosity (and because I'm probably going to use the cleanup method next week 😄) ... what's the first use-case you have in mind for cleanup options?

We have a sql storage mechanism internally that needs this in order to be interruptible for infrastructure stability reasons.

File: test/experiment_test.rb [hunk 1]
```diff
@@ -450,4 +459,10 @@ def test_cleanup_without_redis
       experiment.cleanup
     end
   end
+
+  private
+
+  def redis
+    @redis ||= ::Redis.new(host: REDIS_HOST, port: REDIS_PORT)
```
Comment (BoGs) [inline test/experiment_test.rb]: what redis is this connecting to?

Comment (pawandubey) [reply test/experiment_test.rb]: Some local instance. Previously, I've had to explicitly start one using `redis-server` to make redis dependent tests pass locally.

Comment (gmcgibbon) [reply test/experiment_test.rb]: Correct, the default one is running on localhost at the default port. This works on CI and locally when running redis-server, but redis setup is like this on master.

**Pull request merged**